util/gem5-resources-manager-v2/src/gem5_resources_manager/creator/creator.py

```diff
@@ -0,0 +1,154 @@
+        self.data_source = None
+
+    # Implement the two abstract classes declared in AbstractSubtool.
+    def get_arg_parser(self, subparser):
+        parser = subparser.add_parser("gem5-resources-creator")
+        parser.add_argument("category", type=str, help="category of resource to create")
+        parser.add_argument(
+            "--output",
+            "-o",
+            type=str,
+            help="output file",
+            default="resources.json",
+        )
+        parser.add_argument(
+            "--ignore-schema-validation",
+            "-i",
+            action="store_true",
+            help="ignore validation of resource against schema",
+        )
+        parser.add_argument(
+            "--ignore-db-check",
+            "-d",
+            action="store_true",
+            help="ignore the check that the entered resource does not exist in the database.",
+        )
+        parser.add_argument(
+            "--handle-url",
+            "-u",
+            action="store_true",
+            help="automatically fill in the is_zipped, is_tarred, and md5sum fields if the url field has a link to the source repository.",
+        )
+        parser.add_argument(
+            "--verbose",
+            "-v",
+            action="store_true",
+            help="show more information about the fields",
+        )
+        parser.add_argument(
+            "--required-fields-only",
+            "-r",
+            action="store_false",
+            help="only show required fields when entering data",
+        )
+        parser.add_argument(
+            "--field-entries",
+            "-f",
+            type=str,
+            help="json file dictionary containing field entries",
+            default="{}",
+        )
+
+    def execute(self, args: Namespace, data_source: AbstractDataSource):
+        self.data_source = data_source
+        print("Creator.execute()", args.category)
+        self.create_resource_json(args)
+
+    def create_resource_json(self, args: Namespace):
+        resource = {}
+        required, optional = self.data_source.get_fields(args.category)
+        populated_fields = ast.literal_eval(args.field_entries)
+        resource["category"] = args.category
+        del required["category"]
+
+        self.enter_fields(required, resource, populated_fields, args)
+        if args.required_fields_only:
+            self.enter_fields(
+                optional, resource, populated_fields, args, is_optional=True
+            )
+
+        if not args.ignore_db_check and self.data_source.resource_exists(
+            resource_id=resource["id"]
+        ):
+            # throw runtime exception
+            raise Exception("Resource already exists in database")
+
+        if not args.ignore_schema_validation:
+            self.data_source.validate_resource(resource)
+        self.data_source.create_new_entry(resource)
+
+    def enter_fields(self, fields, resource, populated_fields, args, is_optional=False):
```

Comment:
There's so much confusion here. What information do we want? I think

* The list of fiends for each category.
* For each field a to know if ti's optonal.
* The "type" fo the field (its "schema").
* If we're running in `verbose` mode.
* If `is_optional`.

So, how about you create a new class called `FieldType` which encapulates all the schema information about a field. This woulde include info on whether its optinal, andprovide the field's accepted value types as well as a validator. We can then have a function which returns a List of `Field` objects for that category

Also, let's rename `is_optional` to `ask_on_optionmal_param`. `is_optional` doesn't have any meaning to me.

 Here's the basic pseudo-coee:



```py
field_types = get_field_types(category="binary")

for field_type in field_types:
    if field_type.is_optional():
        if not ask_on_optional_param:
            continue # No inclusion of optional parameters.
        # Here you do "choice logic" asking if they want to 
        # set the parameter. If "no" `continue` called.
    
    val = None
    while not val:
        print(f"The field {field_type.get_name()} expects a value of {field_type.get_value_type_as_str()}")
        val = input('bla bla')
        if not field_type.is_valid_value(val):
            val = None
            print("Incorrect value type!")
 
    # Set the value to the resource here
```

And we've discussed thia many, many times now, and I've still got no movement on it: Can we please cleanup this default value/None value/field missing mess. I really think it'd help a lot if we implemented the following rule:

1. A missing field is consider to have a value of `None`.
2. In line with 1. an "Optional" field is a field that can be `None` or a value.
3. We drop this default value thing. If a field is optional and not set by the user, it's stored as None. If it's not optional then the user is promted for a value which must be given, with this value being valid in accordance to the schema.

Reply:
Here is my interpretation of this:
- If a field is missing that is the key does not exist then it means that the resource doesnot have the field
- Currently the fields cannot be None or Null but if they are lets say "" or [] it means that the resource has the field but the value is an empty string or empty array.